File: android/HapticsPlugin.kt
package dev.wefter.bridge

import android.content.Context
import android.os.Build
import android.os.VibrationEffect
import android.os.Vibrator
import android.os.VibratorManager
import org.json.JSONObject

class HapticsPlugin(context: Context, dispatcher: BridgeDispatcher) :
        WefterPlugin(context, dispatcher) {

    private val vibrator: Vibrator by lazy {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            val manager =
                    context.getSystemService(Context.VIBRATOR_MANAGER_SERVICE) as VibratorManager
            manager.defaultVibrator
        } else {
            @Suppress("DEPRECATION") context.getSystemService(Context.VIBRATOR_SERVICE) as Vibrator
        }
    }

    @WefterMethod
    fun isAvailable(payload: JSONObject, callback: (Result<Any>) -> Unit) {
        val hasVibrator = vibrator.hasVibrator()
        val amplitudeControlSupported =
                hasVibrator &&
                        Build.VERSION.SDK_INT >= Build.VERSION_CODES.O &&
                        vibrator.hasAmplitudeControl()
        resolve(
                callback,
                JSONObject()
                        .put("available", hasVibrator)
                        .put("amplitudeControlSupported", amplitudeControlSupported)
        )
    }

    @WefterMethod
    fun impact(payload: JSONObject, callback: (Result<Any>) -> Unit) {
        val style = payload.optString("style", "medium")
        val preset = IMPACT_PRESETS[style]
        if (preset == null) {
            reject(
                    callback,
                    "INVALID_STYLE",
                    "Unknown impact style: $style. Valid styles are: ${IMPACT_PRESETS.keys.joinToString(", ")}."
            )
            return
        }

        val intensity =
                if (payload.has("intensity") && !payload.isNull("intensity")) {
                    payload.optDouble("intensity")
                } else {
                    null
                }
        val amplitude = intensity?.let { toAmplitude(it) } ?: preset.amplitude

        val predefinedEffect =
                if (intensity == null && Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
                    when (style) {
                        "light" -> VibrationEffect.EFFECT_TICK
                        "medium" -> VibrationEffect.EFFECT_CLICK
                        "heavy" -> VibrationEffect.EFFECT_HEAVY_CLICK
                        else -> null
                    }
                } else {
                    null
                }

        triggerVibration(
                callback,
                duration = preset.durationMs,
                amplitude = amplitude,
                predefinedEffect = predefinedEffect
        )
    }

    @WefterMethod
    fun notification(payload: JSONObject, callback: (Result<Any>) -> Unit) {
        when (payload.optString("type", "")) {
            "success" ->
                    triggerVibration(
                            callback,
                            pattern = longArrayOf(0, 40, 60, 40),
                            amplitude = 180
                    )
            "warning" -> triggerVibration(callback, duration = 60, amplitude = 200)
            "error" ->
                    triggerVibration(
                            callback,
                            pattern = longArrayOf(0, 50, 80, 50, 80, 50),
                            amplitude = 255
                    )
            else -> {
                val type = payload.optString("type", "")
                reject(
                        callback,
                        "INVALID_TYPE",
                        "Unknown notification type: $type. Valid types are: success, warning, error."
                )
            }
        }
    }

    @WefterMethod
    fun selection(payload: JSONObject, callback: (Result<Any>) -> Unit) {
        val predefinedEffect =
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) VibrationEffect.EFFECT_TICK
                else null
        triggerVibration(
                callback,
                duration = 8,
                amplitude = 80,
                predefinedEffect = predefinedEffect
        )
    }

    @WefterMethod
    fun vibrate(payload: JSONObject, callback: (Result<Any>) -> Unit) {
        val amplitude =
                if (payload.has("amplitude") && !payload.isNull("amplitude")) {
                    toAmplitude(payload.optDouble("amplitude"))
                } else {
                    VibrationEffect.DEFAULT_AMPLITUDE
                }

        val patternArray = payload.optJSONArray("pattern")
        if (patternArray != null) {
            if (patternArray.length() == 0) {
                reject(callback, "INVALID_PATTERN", "pattern must not be empty")
                return
            }

            val jsTimings = LongArray(patternArray.length()) { i -> patternArray.optLong(i, 0) }
            if (jsTimings.any { it < 0 }) {
                reject(callback, "INVALID_PATTERN", "pattern durations must not be negative")
                return
            }

            val androidTimings = LongArray(jsTimings.size + 1)
            jsTimings.copyInto(androidTimings, destinationOffset = 1)

            val repeatFlag = payload.optBoolean("repeat", false)
            triggerVibration(
                    callback,
                    pattern = androidTimings,
                    amplitude = amplitude,
                    repeat = repeatFlag
            )
            return
        }

        val duration = payload.optLong("duration", 200L)
        if (duration <= 0) {
            reject(callback, "INVALID_DURATION", "duration must be greater than 0")
            return
        }
        triggerVibration(callback, duration = duration, amplitude = amplitude)
    }

    @WefterMethod
    fun cancel(payload: JSONObject, callback: (Result<Any>) -> Unit) {
        vibrator.cancel()
        resolve(callback, JSONObject().put("cancelled", true))
    }

    private fun triggerVibration(
            callback: (Result<Any>) -> Unit,
            duration: Long? = null,
            pattern: LongArray? = null,
            amplitude: Int = VibrationEffect.DEFAULT_AMPLITUDE,
            repeat: Boolean = false,
            predefinedEffect: Int? = null,
    ) {
        if (!vibrator.hasVibrator()) {
            reject(callback, "NOT_AVAILABLE", "This device has no vibrator")
            return
        }

        try {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                val effect =
                        when {
                            predefinedEffect != null &&
                                    Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q ->
                                    VibrationEffect.createPredefined(predefinedEffect)
                            pattern != null ->
                                    VibrationEffect.createWaveform(
                                            pattern,
                                            IntArray(pattern.size) { i ->
                                                if (i % 2 == 0) 0 else amplitude
                                            },
                                            if (repeat) 0 else -1
                                    )
                            else ->
                                    VibrationEffect.createOneShot(
                                            (duration ?: 200L).coerceAtLeast(1),
                                            amplitude
                                    )
                        }
                vibrator.vibrate(effect)
            } else {
                @Suppress("DEPRECATION")
                if (pattern != null) {
                    vibrator.vibrate(pattern, if (repeat) 0 else -1)
                } else {
                    vibrator.vibrate((duration ?: 200L).coerceAtLeast(1))
                }
            }
            resolve(callback, JSONObject().put("played", true))
        } catch (e: IllegalArgumentException) {
            reject(callback, "INVALID_PATTERN", e.message ?: "Invalid vibration pattern")
        } catch (e: Exception) {
            reject(callback, "VIBRATE_FAILED", e.message ?: "Could not trigger vibration")
        }
    }

    private fun toAmplitude(intensity: Double): Int =
            (intensity.coerceIn(0.0, 1.0) * 254 + 1).toInt().coerceIn(1, 255)

    private data class ImpactPreset(val durationMs: Long, val amplitude: Int)

    companion object {
        private val IMPACT_PRESETS =
                mapOf(
                        "light" to ImpactPreset(10L, 90),
                        "soft" to ImpactPreset(15L, 110),
                        "medium" to ImpactPreset(20L, 150),
                        "rigid" to ImpactPreset(15L, 200),
                        "heavy" to ImpactPreset(30L, 255),
                )
    }
}

File: ios/HapticsPlugin.swift
import CoreHaptics
import UIKit

final class HapticsPlugin: WefterPlugin {

    private var hapticEngine: CHHapticEngine?
    private var activePlayer: CHHapticPatternPlayer?

    // @WefterMethod
    func isAvailable(payload: [String: Any], callback: @escaping (Result<Any, Error>) -> Void) throws {
        let supportsHaptics = CHHapticEngine.capabilitiesForHardware().supportsHaptics
        resolve(callback, data: [
            "available": supportsHaptics,
            "amplitudeControlSupported": supportsHaptics,
        ])
    }

    // @WefterMethod
    func impact(payload: [String: Any], callback: @escaping (Result<Any, Error>) -> Void) throws {
        let styleName = payload["style"] as? String ?? "medium"
        guard let style = HapticsPlugin.impactStyle(for: styleName) else {
            reject(callback, code: "INVALID_STYLE", message: "Unknown impact style: \(styleName). Valid styles are: light, soft, medium, rigid, heavy.")
            return
        }

        let generator = UIImpactFeedbackGenerator(style: style)
        generator.prepare()
        if let intensity = (payload["intensity"] as? NSNumber)?.doubleValue {
            generator.impactOccurred(intensity: CGFloat(min(max(intensity, 0), 1)))
        } else {
            generator.impactOccurred()
        }
        resolve(callback, data: ["played": true])
    }

    // @WefterMethod
    func notification(payload: [String: Any], callback: @escaping (Result<Any, Error>) -> Void) throws {
        let typeName = payload["type"] as? String ?? ""
        guard let type = HapticsPlugin.notificationType(for: typeName) else {
            reject(callback, code: "INVALID_TYPE", message: "Unknown notification type: \(typeName). Valid types are: success, warning, error.")
            return
        }

        let generator = UINotificationFeedbackGenerator()
        generator.prepare()
        generator.notificationOccurred(type)
        resolve(callback, data: ["played": true])
    }

    // @WefterMethod
    func selection(payload: [String: Any], callback: @escaping (Result<Any, Error>) -> Void) throws {
        let generator = UISelectionFeedbackGenerator()
        generator.prepare()
        generator.selectionChanged()
        resolve(callback, data: ["played": true])
    }

    // @WefterMethod
    func vibrate(payload: [String: Any], callback: @escaping (Result<Any, Error>) -> Void) throws {
        let rawAmplitude = (payload["amplitude"] as? NSNumber)?.doubleValue ?? 1.0
        let amplitude = Float(min(max(rawAmplitude, 0), 1))

        if let patternArray = payload["pattern"] as? [NSNumber] {
            guard !patternArray.isEmpty else {
                reject(callback, code: "INVALID_PATTERN", message: "pattern must not be empty")
                return
            }

            let timingsMs = patternArray.map { $0.doubleValue }
            guard timingsMs.allSatisfy({ $0 >= 0 }) else {
                reject(callback, code: "INVALID_PATTERN", message: "pattern durations must not be negative")
                return
            }

            let repeatFlag = payload["repeat"] as? Bool ?? false
            playPattern(timingsSeconds: timingsMs.map { $0 / 1000.0 }, amplitude: amplitude, repeats: repeatFlag, callback: callback)
            return
        }

        let durationMs = (payload["duration"] as? NSNumber)?.doubleValue ?? 200
        guard durationMs > 0 else {
            reject(callback, code: "INVALID_DURATION", message: "duration must be greater than 0")
            return
        }
        playPattern(timingsSeconds: [durationMs / 1000.0], amplitude: amplitude, repeats: false, callback: callback)
    }

    // @WefterMethod
    func cancel(payload: [String: Any], callback: @escaping (Result<Any, Error>) -> Void) throws {
        if let player = activePlayer {
            try? player.stop(atTime: CHHapticTimeImmediate)
            activePlayer = nil
        }
        resolve(callback, data: ["cancelled": true])
    }

    private func playPattern(timingsSeconds: [Double], amplitude: Float, repeats: Bool, callback: @escaping (Result<Any, Error>) -> Void) {
        guard CHHapticEngine.capabilitiesForHardware().supportsHaptics else {
            reject(callback, code: "NOT_AVAILABLE", message: "This device does not support custom haptic patterns")
            return
        }

        do {
            let engine = try activeEngine()

            var events: [CHHapticEvent] = []
            var time: Double = 0
            for (index, segment) in timingsSeconds.enumerated() {
                if index % 2 == 0 {
                    let intensity = CHHapticEventParameter(parameterID: .hapticIntensity, value: amplitude)
                    let sharpness = CHHapticEventParameter(parameterID: .hapticSharpness, value: 0.5)
                    events.append(CHHapticEvent(eventType: .hapticContinuous, parameters: [intensity, sharpness], relativeTime: time, duration: segment))
                }
                time += segment
            }

            guard !events.isEmpty else {
                reject(callback, code: "INVALID_PATTERN", message: "pattern must include at least one on-segment")
                return
            }

            let pattern = try CHHapticPattern(events: events, parameterCurves: [])

            if repeats {
                let player = try engine.makeAdvancedPlayer(with: pattern)
                player.loopEnabled = true
                try player.start(atTime: CHHapticTimeImmediate)
                activePlayer = player
            } else {
                let player = try engine.makePlayer(with: pattern)
                try player.start(atTime: CHHapticTimeImmediate)
                activePlayer = player
            }

            resolve(callback, data: ["played": true])
        } catch {
            reject(callback, code: "VIBRATE_FAILED", message: "Could not trigger vibration: \(error.localizedDescription)")
        }
    }

    private func activeEngine() throws -> CHHapticEngine {
        if let engine = hapticEngine {
            return engine
        }

        let engine = try CHHapticEngine()
        engine.stoppedHandler = { [weak engine] _ in
            try? engine?.start()
        }
        engine.resetHandler = { [weak engine] in
            try? engine?.start()
        }
        try engine.start()
        hapticEngine = engine
        return engine
    }

    private static func impactStyle(for name: String) -> UIImpactFeedbackGenerator.FeedbackStyle? {
        switch name {
        case "light": return .light
        case "soft": return .soft
        case "medium": return .medium
        case "rigid": return .rigid
        case "heavy": return .heavy
        default: return nil
        }
    }

    private static func notificationType(for name: String) -> UINotificationFeedbackGenerator.FeedbackType? {
        switch name {
        case "success": return .success
        case "warning": return .warning
        case "error": return .error
        default: return nil
        }
    }
}
